panel/dataclass.py

```diff
@@ -0,0 +1,507 @@
+    Returns a Viewer object from a model with parameters synced
+    bidirectionally to the model's fields and displaying the model
+    when rendered.
+
+    Arguments
+    ---------
+    model: DataClassLike
+        The model to create the Viewer from.
+    names: Iterable[str] | dict[str, str]
+        The names of the parameters to add to the Viewer and to sync bidirectionally.
+        If no names are specified, all public and relevant fields on the model will be added and synced.
+    bases: tuple[type]
+        Additional base classes to add to the base `ModelViewer` class.
+    kwargs: dict[str, Any]
+        Additional keyword arguments to pass to the Parameterized constructor.
+
+    Returns
+    -------
+    A ModelViewer instance wrapping the supplied model.
+    """
+    bases = _to_tuple(bases) + (ModelViewer,)
+    return to_parameterized(model, names, bases=bases, **kwargs)
+
+
+def sync_with_rx(model: HasTraits, name: str, rx: param.rx):
+    """
+    Syncs a single field of a model with an `rx` value.
+
+    Arguments
+    ---------
+    model: HasTraits | pydantic.Model
+        The model to sync with.
+    name: str
+        The name of the field to sync the `rx` value with.
+    rx: param.rx
+        The `rx` value to sync with the field.
+    """
+    rx.rx.value = getattr(model, name)
+
+    def set_value(event, rx=rx):
+        rx.rx.value = event["new"]
+
+    _get_utils(model).observe_field(model, name, set_value)
+
+    def set_name(value, element=model, name=name):
+        setattr(element, name, value)
+
+    rx.rx.watch(set_name)
+
+
+def to_rx(model: HasTraits, *names) -> param.rx | tuple[param.rx]:
+    """
+    Returns `rx` values from fields of a model, each synced to a field of the model bidirectionally.
+
+    Arguments
+    ---------
+    model: HasTraits | pydantic.BaseModel
+        The model to create the `rx` parameters from.
+    names: Iterable[str]
+        The fields to create `rx` values from.
+        If a single parameter is specified, a single reactive parameter is returned.
+        If no names are specified, all public and relevant fields of the model will be used.
+
+    Returns
+    -------
+    One or more `rx` objects corresponding to the fields of the object.
+
+    Example
+    -------
+
+    ```python
+    import panel as pn
+    import ipyleaflet as ipyl
+
+    pn.extension("ipywidgets")
+
+    leaflet_map = ipyl.Map(zoom=4)
+    zoom, zoom_control = pn.dataclass.to_rx(
+        leaflet_map, "zoom", "zoom_control"
+    )
+
+    pn.Column(
+        leaflet_map,
+        pn.rx("**Value**: {zoom}, **Zoom Control**: {zoom_control}").format(
+           zoom=zoom, zoom_control=zoom_control
+        ),
+    ).servable()
+    ```
+    """
+    rx_values = []
+    for name in names:
+        rx = param.rx()
+        sync_with_rx(model, name, rx)
+        rx_values.append(rx)
+    if len(rx_values) == 1:
+        return rx_values[0]
+    return tuple(rx_values)
+
+
+class ModelForm(Viewer):
```

Comment:
I added the `ModelForm` to support the *form* use case. This is the use case that pydantic-panel, streamlit-pydantic and dash-pydantic-form all support.

Its really a more general request as in https://github.com/holoviz/panel/issues/3687. We can solve this is several ways:

- not solve :-)
- document how to solve
- make general Panel form functionality
- add this dataclass specific form functionality.

What do you recommend @philippjfr ?

Reply:
Its worth noting that I achieve the *form* functionality a bit differently than the packages mentioned above. They translate pydantic fields to widgets while I translate pydantic fields to Parameters which Panel can translate to widgets. I.e. in my version we instantiate a Parameterized and then use Param. In the packages they can avoid instantiating a model before the user has filled out the form including required values.

Is my version the right way to go @philippjfr ?

Reply:
Great work! And the form use case should definitely be supported.